Core/MIPS/MIPSIntVFPU.cpp

```diff
@@ -348,80 +348,70 @@ namespace MIPSInt
 	{
 		int vd = _VD;
 		VectorSize sz = GetVecSize(op);
-		int n = GetNumVectorElements(sz);
-		static const float ones[4] = {1,1,1,1};
-		static const float zeros[4] = {0,0,0,0};
-		const float *v;
-		switch ((op >> 16) & 0xF)
-		{
-		case 6: v=zeros; break;  //vzero
-		case 7: v=ones; break;   //vone
+		float d[4];
+
+		int constant = 0;
+		switch ((op >> 16) & 0xF) {
+		case 6: constant = 0; break;  //vzero
+		case 7: constant = 1; break;   //vone
 		default:
-			_dbg_assert_msg_(CPU,0,"Trying to interpret instruction that can't be interpreted");
+			_dbg_assert_msg_(CPU, 0, "Trying to interpret instruction that can't be interpreted");
 			PC += 4;
 			EatPrefixes();
 			return;
 		}
-		float o[4];
-		for (int i = 0; i < n; i++)
-			o[i] = v[i];
-		ApplyPrefixD(o, sz);
-		WriteVector(o, sz, vd);
+
+		// The S prefix generates constants, but negate is still respected.
+		u32 sprefixRemove = VFPU_ANY_SWIZZLE();
+		u32 sprefixAdd = VFPU_CONST(1, 1, 1, 1) | VFPU_SWIZZLE(constant, constant, constant, constant);
+		ApplyPrefixST(d, VFPURewritePrefix(VFPU_CTRL_SPREFIX, sprefixRemove, sprefixAdd), sz);
+
+		ApplyPrefixD(d, sz);
+		WriteVector(d, sz, vd);
 
 		EatPrefixes();
 		PC += 4;
 	}
 
-	void Int_Viim(MIPSOpcode op)
-	{
+	void Int_Viim(MIPSOpcode op) {
 		int vt = _VT;
 		s32 imm = (s16)(op&0xFFFF);
 		u16 uimm16 = (op&0xFFFF);
-		//V(vt) = (float)imm;
 		float f[1];
 		int type = (op >> 23) & 7;
-		if (type == 6)
+		if (type == 6) {
 			f[0] = (float)imm;  // viim
-		else if (type == 7)
+		} else if (type == 7) {
 			f[0] = Float16ToFloat32((u16)uimm16);   // vfim
-		else
-		{
+		} else {
 			_dbg_assert_msg_(CPU,0,"Trying to interpret instruction that can't be interpreted");
 			f[0] = 0;
 		}
 		
 		ApplyPrefixD(f, V_Single);
-		V(vt) = f[0];
+		WriteVector(f, V_Single, vt);
 		PC += 4;
 		EatPrefixes();
 	}
 
-	void Int_Vidt(MIPSOpcode op)
-	{
+	void Int_Vidt(MIPSOpcode op) {
 		int vd = _VD;
 		VectorSize sz = GetVecSize(op);
 		float f[4];
-		switch (sz)
-		{
-		case V_Pair:
-			f[0] = (vd&1)==0 ? 1.0f : 0.0f;
-			f[1] = (vd&1)==1 ? 1.0f : 0.0f;
-			break;
-		case V_Quad:
-			f[0] = (vd&3)==0 ? 1.0f : 0.0f;
-			f[1] = (vd&3)==1 ? 1.0f : 0.0f;
-			f[2] = (vd&3)==2 ? 1.0f : 0.0f;
-			f[3] = (vd&3)==3 ? 1.0f : 0.0f;
-			break;
-		default:
-			_dbg_assert_msg_(CPU,0,"Trying to interpret instruction that can't be interpreted");
-			break;
-		}
+
+		// The S prefix generates constants, but negate is still respected.
+		int offmask = sz == V_Quad || sz == V_Triple ? 3 : 1;
+		int off = vd & offmask;
+		u32 sprefixRemove = VFPU_ANY_SWIZZLE();
+		u32 sprefixAdd = VFPU_CONST(1, 1, 1, 1) | VFPU_SWIZZLE(off == 0, off == 1, off == (2 & offmask), off == (3 & offmask));
```

Comment:
I have to admit I'm a little bit confused here. We generate four ones, and swizzle them?

Reply:
This is for identity.  I could make the macros more clear, though.

`VFPU_CONST(1, 1, 1, 1)` just means set the constant flag to 1 for each of the lanes.  It doesn't specify the value of the constant.

`VFPU_SWIZZLE(...)` and `VFPU_ABS()` values are actually used to select the constant (when the flag is set.)

I'll create a separate macro for the const values.

-[Unknown]

Reply:
Updated, I think I would've had trouble understanding this code too in 6 months, the macros should make it clearer (used a separate commit since it changes other ops that already used the old way.)

-[Unknown]

Reply:
Oh duh, I forgot that the CONST flag just switched the mode. I appreciate the new macros, they indeed make things a lot more clearer.